---
title: "Project Bioinformatic Resources"
author: "Annalisa Xamin, Nicola Perotti"
output:
  pdf_document:
    latex_engine: xelatex
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE,
                      message=FALSE,
                      tidy.opts=list(width.cutoff = 80),
                      tidy = TRUE,
                      fig.align = "center"
                      )

# Set working directory
setwd("~/Documents/Repo_Git/Project_BR")
```


# Introduction
The following analysis focus on RNA-seq count data extracted from different cancer datasets from the Cancer Genome Atlas (TCGA). From the original TCGA data 50 cases (tumor samples) and 50 controls (normal samples) were randomly selected.

# Analysis
The packages needed for the analysis are the following:
```{r load libraries, results='hide'}
library(sessioninfo)
library(tidyverse)
library(biomaRt)
library(MotifDb) # large collection of motifs across different species
library(seqLogo) 
library(PWMEnrich)
library(PWMEnrich.Hsapiens.background)
library(GEOquery)
library(oligo)
library(pd.hg.u133.plus.2)
library(hgu133plus2.db)
library(genefilter)
library(limma)
library(pheatmap)
library(stringr)
library(GenomicFeatures) # to build object with specific information such as genomic coordinates
library(ggplot2)
library(edgeR) # designed to perform Differential Expression Analysis from RNA-Seq data
library(fgsea) # computation of enrichment scores and other statistics
library(org.Hs.eg.db) # database annotation human specific transcript from Ensembl
library(clusterProfiler) # uses Fisher test, explore gene list of interest against a reference
library(enrichplot) # build enrich map
library(ggnewscale)
library(DOSE)
library(pathview) # contains cartoons of pathways and we project on them our genes of interest
```

## Load data
In particular, we consider data coming from **Lung adenocarcinoma**.
```{r Load data}
load("./RData/Lung_adenocarcinoma.RData")
```
After loading the data, the following three data-frames are available:

- `raw_counts_df` which contains the raw RNA-seq counts;

- `c_anno_df`, which contains sample name and condition (case and control);

- `r_anno_df`, which contains the ENSEMBL genes ids, the length of the genes and the genes symbols.

```{r explore raw_counts_df, include=FALSE}
# Explore only the first 5 rows of the data frame
head(raw_counts_df, n=5)
head(rownames(raw_counts_df), n=5)
head(colnames(raw_counts_df), n=5)
```

```{r explore c_anno_df, include = FALSE}
# Explore only the first 5 rows of the data frame
head(c_anno_df, n = 5)
head(rownames(c_anno_df), n=5)
head(colnames(c_anno_df)) # there are only 2 columns: sample and condition
```

```{r explore r_anno_df, include = FALSE}
# Explore only the first 5 rows of the data frame
head(r_anno_df, n = 5)
head(rownames(r_anno_df), n=5)
head(colnames(r_anno_df)) # there are only 3 columns: ensembl_gene_id, external_gene_name, length
```

```{r}
# rownames(c_anno_df) <- c_anno_df$sample
# c_anno_add <- as.data.frame(str_split(string=c_anno_df$sample, pattern="_", simplify = T))
# colnames(c_anno_add) <- c("condition","replicate")
# c_anno_df <- cbind(c_anno_df,c_anno_add)
# c_anno_df
```

## Filter protein coding genes
To extract only protein coding genes from `raw_counts_df` and `r_anno_df`, we use the `biomaRt` package.

```{r attributes, include = FALSE}
# get the list of possible attributes
ensembl <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
attributes <- listAttributes(ensembl)
```
First, we retrieve the information about the protein coding genes from Ensembl.
```{r extract protein coding genes}
ensembl <- useDataset("hsapiens_gene_ensembl",mart=ensembl)
query <- getBM(attributes=c("ensembl_gene_id",
                            "external_gene_name",
                            "gene_biotype"),
                filters = c("ensembl_gene_id"), 
                values = r_anno_df$ensembl_gene_id,
                mart = ensembl)

query_protein_coding <- query[which(query$gene_biotype == "protein_coding"), ]

```

```{r inspect query_protein_coding, include=FALSE}
query_protein_coding
```

Then, we filter the data frames containing the raw counts and the annotation to keep only the protein coding genes.
```{r raw_counts r_anno df protein coding}
indexes_r_anno_df <- which(r_anno_df$ensembl_gene_id %in% query_protein_coding$ensembl_gene_id)
r_anno_df_protein_coding <- r_anno_df[indexes_r_anno_df, ]

indexes_raw_counts_df <- which(rownames(raw_counts_df) %in% query_protein_coding$ensembl_gene_id)
raw_counts_df_protein_coding <- raw_counts_df[indexes_raw_counts_df, ]
```

## Differential expression analysis
To perform the differential expression analysis, we will use the `edgeR` package. 

It is important to remove genes with low signal that will have low statistical power. Since, we want to focus on the transcripts we can use to perform the analysis, we can filter raw counts data using a threshold of raw count > 20 in at least 5 replicates.
```{r filtering}
# count threshold
count_thr <- 20

# number of replicates with more counts than the count threshold
repl_thr <- 5

filter_vec <- apply(raw_counts_df_protein_coding,1, # go through all count matrices by rows
    function(y) max(by(y, c_anno_df$condition, function(x) sum(x>=count_thr))))
    # groups the values on each condition and sum all the values above the count
```

```{r see statistics for the filtering, include=FALSE}
table(filter_vec)
```
Then, we filter the previously updated data frames.
```{r create filtered df }
filter_counts_df <- raw_counts_df_protein_coding[filter_vec>=repl_thr,]

filter_anno_df <- r_anno_df_protein_coding[rownames(filter_counts_df),]
```

```{r TMM normalization}
edge_c <- DGEList(counts=filter_counts_df,group=c_anno_df$condition,samples=c_anno_df,genes=filter_anno_df)

# computing norm factors for TMM normalization
edge_n <- calcNormFactors(edge_c,method="TMM")
```

```{r cpm_table}
# create a cpm table (normalized expression values)
cpm_table <- as.data.frame(round(cpm(edge_n),2))
```

```{r, include=FALSE}
long_cpm_df <- gather(cpm_table, key = "sample", value = "CPM")

ggplot(data=long_cpm_df,aes(sample,CPM+1)) +
   geom_boxplot(colour="olivedrab",fill="olivedrab",alpha=0.7)+
   theme_bw()+
   scale_y_log10() 
```

```{r}
design <- model.matrix(~0+group, data=edge_n$samples)
colnames(design) <- levels(edge_n$samples$group)
rownames(design) <- edge_n$samples$sample
```

```{r insepct design, include=FALSE}
design
```


```{r dispersion and fit}
# calculate dispersion and fit with edgeR (necessary for differential expression analysis)
edge_d <- estimateDisp(edge_n,design)
edge_f <- glmQLFit(edge_d,design) 
```

```{r define contrasts}
contro <- makeContrasts("control-case", levels=design)
```
`estimateDisp` $$\rightarrow$$ The tool tries to estimate the variability at different levels: in the data, inter sample and intra sample.

```{r DE}
# fit the model with generalized linear models
edge_t <- glmQLFTest(edge_f,contrast=contro)
DEGs <- as.data.frame(topTags(edge_t, n=20,p.value = 0.01,sort.by = "logFC")) # returns object with best 20 top genes
DEGs <- as.data.frame(topTags(edge_t,n=20000)) # select all the 20000 genes

# We can improve the selection, considering not only the p-value, but also
# the average expression of the genes (logCPM).
DEGs$class <- "="
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC>1.5&DEGs$FDR<0.25)] = "+" # upregulated genes
DEGs$class[which(DEGs$logCPM>1&DEGs$logFC<(-1.5)&DEGs$FDR<0.25)] = "-" # downregulated genes
DEGs <- DEGs[order(DEGs$logFC,decreasing = T),]
```

```{r check how many up and down regulated}
head(DEGs)
table(DEGs$class) # 1202-, 885+, 15202= 
```

```{r volcano plot}
input_df <- DEGs
xlabel <- "log2 FC case vs control"
ylabel <- "-log10 p-value"

par(fig=c(0,1,0,1), mar=c(4,4,1,2), mgp=c(2, 0.75, 0))	
plot(input_df$logFC, -log(input_df$PValue,base=10),xlab=xlabel, ylab=ylabel, 
     col=ifelse(input_df$class=="=","grey70","olivedrab4"), pch=20, frame.plot=TRUE, cex=0.8, main="Volcano plot")
abline(v=0,lty=2,col="grey20")
```

```{r heatmap}
# only up and down regulated genes
cols <- c(ifelse(c_anno_df$condition == "case", "chartreuse4","burlywood3"))
pal <- c("blue","white","red") 
pal <- colorRampPalette(pal)(100)
heatmap(as.matrix(cpm_table[which(rownames(cpm_table)%in%DEGs$ensembl_gene_id[which(DEGs$class!="=")]),]),
        ColSideColors = cols,cexCol = 0.5,margins = c(4,4),col=pal,cexRow = 0.2)
```

```{r save up and down regulated in text file}
up_DEGs <- DEGs[which(DEGs$class=="+"),]
down_DEGs <- DEGs[which(DEGs$class=="-"),]

write.table(up_DEGs,file="up_DEGs.txt",row.names=F,col.names=T,sep="\t",quote=F)
write.table(down_DEGs,file="down_DEGs.txt",row.names=F,col.names=T,sep="\t",quote=F)
write.table(DEGs,file="DEGs.txt",row.names=F,col.names=T,sep="\t",quote=F)
```


## Gene set enrichment analysis
To perform the gene set enrichment analysis, we will use the `clusterProfiler` package.

```{r import the previously saved DEGs}
DEGs <- read.table("DEGs.txt",header=T,sep="\t",as.is=T)
table(DEGs$class)
```


```{r}
ensembl <- useEnsembl(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
convert <- getBM(attributes=c("ensembl_gene_id","entrezgene_id","external_gene_name"),
                 filters=c("ensembl_gene_id"), 
                 values=DEGs$ensembl_gene_id,
                 mart = ensembl)

DEGs <- merge(DEGs,convert,by.x="ensembl_gene_id",by.y="ensembl_gene_id") # include the new info in the original data frame
DEGs <- DEGs[which(!is.na(DEGs$entrezgene_id)),]
DEGs <- DEGs[-which(duplicated(DEGs$entrezgene_id)),]
```


```{r rank, include= FALSE}
ranks <- DEGs$logFC
ranks <- sort(ranks, decreasing = T) # sort manually to have the order needed for the package
names(ranks) <- DEGs$entrezgene_id
head(ranks)
barplot(sort(ranks, decreasing = T),las=3)
```


```{r}
# list of up-regulated genes
upDEGs <- DEGs %>% filter(class == "+")
# list of down-regulated genes
downDEGs <- DEGs %>% filter(class == "-")
```

```{r BP up regulated}
# biological process up
ego_BP <- enrichGO(gene = upDEGs$external_gene_name.y,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r BP down regulated}
# biological process down
ego_BP_d <- enrichGO(gene = downDEGs$external_gene_name.y,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "BP",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```


```{r barplot BP up}
barplot(ego_BP,showCategory=10)
```

```{r barplot BP down}
barplot(ego_BP_d,showCategory=10)
```


```{r MP up regulated}
# molecular function up
ego_MF <- enrichGO(gene = upDEGs$external_gene_name.y,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r MP down regulated}
# molecular function down
ego_MF_d <- enrichGO(gene = downDEGs$external_gene_name.y,
                   OrgDb = org.Hs.eg.db,
                   keyType = 'SYMBOL',
                   ont = "MF",
                   pAdjustMethod = "BH",
                   pvalueCutoff = 0.05,
                   qvalueCutoff = 0.05)
```

```{r visualize MF up}
barplot(ego_MF,showCategory=10)
```

```{r visualize MF down}
barplot(ego_MF_d,showCategory=10)
```

```{r WP up regulated}
eWP <- enrichWP(gene = upDEGs$entrezgene_id,
                    organism = 'Homo sapiens',
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.1)

head(eWP, n=10)
```

```{r WP down regulated}
eWP_d <- enrichWP(gene = downDEGs$entrezgene_id,
                    organism = 'Homo sapiens',
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.1)

head(eWP_d, n=10)
```
Chiedere a Romanel per pathway.id (KEGG o WP)?
```{r pathview up regulated}
logFC <- upDEGs$logFC
names(logFC) <- upDEGs$entrezgene_id
pathview(gene.data = logFC, 
         pathway.id = "WP2446", 
         species = "human")

```

```{r KEGG}
eWP_KEGG <- enrichKEGG(gene = upDEGs$entrezgene_id,
                    organism = 'human',
                    pvalueCutoff = 0.05,
                    qvalueCutoff = 0.1)

head(eWP_KEGG, n=20)
```


## Visualization of the enriched pathway


## Enrichment score transcription factors

```{r retrieve promoter regions}
promoter_regions <- getSequence(id=downDEGs$ensembl_gene_id,
                    type="ensembl_gene_id",
                    seqType="gene_flank",
                    upstream = 500,
                    mart=ensembl)
```

```{r}
data(PWMLogn.hg19.MotifDb.Hsap)
seq <- lapply(promoter_regions$gene_flank, function(x) DNAString(x))
```

```{r enriched TFs}
enriched_TFs = motifEnrichment(seq,PWMLogn.hg19.MotifDb.Hsap,score = "affinity")
```

